### PR TITLE
Recipe Diary App Complete

### PR DESCRIPTION
**Pull request by mudasiri**
In this PR myself @mudasiri and @brainskev  did the following:

    A login page.
    A registration page.
    A food list (with all CRUD implementation, except for 'update').
    A recipes list (with all CRUD implementation, except for 'update').
    A general shopping list view (all missing food for all your recipes and total price).
    A list of all public recipes from other users with their names and total prices.

Project requirements

    You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
    Login page and registration page:
        Should be built with Devise.
    Food list (only for a group that has less than 3 members):
        Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
        Should lead to a form that allows users to add new food.
    Recipes list:
        Should display a list of recipes created by the logged-in user as in the wireframe.
        Should lead to recipe details.
        If the user is the owner of the recipe, should allow the user to delete it.
    Public recipe list:
        Should display a list of all public recipes ordered by newest as in the wireframe.
        Should lead to recipe details.
    Recipe details:
        Should display a toggle button that allows for a recipe to be made public or private.
        If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
        If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
    General shopping list view (only for a group that has less than 3 members):
        Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
        Should count the total food items and total price of the missing food.
    Make sure there are no N+1 queries happening.
    Create a navigation menu that allows users to open all of the pages you created.
    Write unit and integration tests.

### CODE REVIEW

File: app/controllers/shopping_lists_controller.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def check_duplicate(foods, food, quantity)
+    foods.each do |food_item|
+      next unless food_item[:name] == food.name
+
+      food_item[:quantity] = food_item[:quantity] - quantity
+      food_item[:price] = food_item[:quantity] * food.price
+      return false
+    end
```
Comment (rloterh) [inline app/controllers/shopping_lists_controller.rb:29]: - [ ] Nice job, team! 👌 However, let's tweak the generation of the shopping list to match the requirements. Let's ensure that lists are generated based on the ingredients added to a particular recipe. Let's also ensure that the estimated quantity and price is derived from multiplying an estimated number by the quantity and price of a corresponding ingredient.

**Your Recipe and Generated Shopping List:**
**Recipe:**        |  **Shopping List:**
:-------------------------:|:-------------------------:
Home Page | Home Page


**Expected:**
image

File: spec/views/recipes_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      expect(page).to have_content(@recipe.description)
+    end
+
+    it 'renders remove button' do
+      expect(page).to have_content('Delete')
+    end
+
+    it 'redirects to delete path' do
+      click_button 'Delete'
+      expect(page).to have_current_path(recipes_path)
```
Comment (rloterh) [inline spec/views/recipes_index_spec.rb]: - [ ] Some of the spec test cases are failing. Let's ensure that all spect test cases are passing!

**Failing test cases:**
image

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Muda & Kelvin, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁 🚦 ⛪🏃 🚜 🏃 🌴 🏃🏫 🌳 🏘️

## Highlights 👍  
- Authentication with Devise gem properly implemented!
- App schema closely follows given ERD!
- Only particular owners can delete contents added by himself/herself!
- All users can view public recipes!
- Inventory properly implemented!
- Unit test framework properly setup!
- Good ReadMe!
- Reviewed projects in turns!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/recipe_foods_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class RecipeFoodsController < ApplicationController
+  def new
+    @recipe_food = RecipeFood.new
+  end
+
+  def create
```
Comment (codecaiine) [inline app/controllers/recipe_foods_controller.rb:6]: - [ ] Good job so far but I noticed an issue with **the generation of the shopping list**. As the previous reviewer mentioned make sure that that lists are generated based on the ingredients added to a particular recipe. Also, make sure that all information about estimated quantity and price is displayed according to the project requirement. 😄 

#### Wireframe
general_shopping_list (1)

#### Your app 

hhhhhh

**Review by codecaiine — CHANGES_REQUESTED** [review]
Hi @mudasiri , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- Good app implementation :heavy_check_mark: 
- Good test implementation :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me (@codecaiine) in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/recipe_foods_controller.rb (hunk 3 above)
Comment (mudasiri) [reply app/controllers/recipe_foods_controller.rb:6]: @codecaiine I don't think that is what the project requirements meant. see it below the requirement said for all recipes not particular recipe as you have stated:
General shopping list view (only for a group that has less than 3 members):
    Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    Should count the total food items and total price of the missing food.

**Review by bobb-Rob — APPROVED** [review]
## STATUS: APPROVED ✔️ 🍾 🍾 🍾 🥳 🥳 🥳 

Hi @mudasiri, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ No Linter error
✅ Correct use of Gitflow
✅ Very descriptive Pull Request(PR)
✅ Neat Commit messages

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**